### DISCUSSION TITLE
Copilot adds extra closing quotes to completions

### DISCUSSION THREAD

**Discussion by wesbos** [community/community#48319] [Copilot Conversations] [answered]
### Select Topic Area

Bug

### Body

Seems dozens of people having this issue: https://twitter.com/wesbos/status/1628458196967751681

I've been experiencing it for about 3-4 weeks.

If I open a quote in VS code, it opens the closing one. Then copilot ignores the closing quote and add it's own.

Example:

<img width="522" alt="Screenshot 2023-02-22 at 3 27 39 PM" src="https://user-images.githubusercontent.com/176013/220751475-18234944-e80c-481f-8270-3f703c64be6b.png">

Any fix for this? Very annoying. Apparently also happening with parenthesis. Will add screenshots as I hit them.

**Comment by calag4n** [comment]
Also with ) , }, ]

**Comment by iamcjwilliams** [comment]
@wesbos Thank you for your feedback, our copilot team has been notified!

**Reply by wesbos** [reply]
awesome! thank you!

**Comment by karolis-kimtys** [comment]
![Screenshot_20230223_071542](https://user-images.githubusercontent.com/46462840/221008343-38ee8fcb-8903-490e-8048-781881668f8a.png)
![Screenshot_20230223_071443](https://user-images.githubusercontent.com/46462840/221008366-024a362f-c059-4c20-bf8e-e24284913bd8.png)
![Screenshot_20230223_071355](https://user-images.githubusercontent.com/46462840/221008373-740c375d-a01f-4095-be67-8090ac0f9fbb.png)
Does not do this all the time, but when it does, it is slightly annoying.

**Comment by johnson-alan** [comment]
Happens intermittently (given the stochastic nature of these LLMs presumably), so my pattern-matching brain can't learn to anticipate it and correct for it: 

<img width="343" alt="Screenshot 2023-02-24 at 7 27 14 AM" src="https://user-images.githubusercontent.com/8865242/221179801-ca557fe2-1dd0-4436-b796-db116b331f13.png">

**Comment by DianomiJH** [comment]
I've also been getting this, an additional case being 
![image](https://user-images.githubusercontent.com/84312830/221195330-8795f2ae-4732-48aa-8ab5-3b752ddcb927.png)
I've already closed the quotes and it's recommended that I close them again

**Comment by NickMaczk** [comment]
I'm getting extra brackets when I validate a suggestion inside brackets.

**Comment by Billycarter1224** [comment]
I am also facing this issue. Please fix it if someone solved it.

**Comment by joehillen** [comment]
https://medium.com/coding-at-dawn/make-github-copilot-way-better-with-these-two-vs-code-settings-842775ba82f7

**Reply by wesbos** [reply]
not a great solution, then we lose this functionality for the entire editor. I hope this is just fixed in co-pilot

**Comment by wesbos** [comment]
~~With the announcement of Github copilot x today, I think this is fixed! I'm on nightly, can anyone confirm for stable?~~

**Reply by KonstantinosPetrakis** [reply]
I am still getting it right now.

**Comment by patosullivan** [comment]
I use copilot.vim, still seeing this issue.

**Comment by wesbos** [comment]
I was wrong, this is still an issue. Even the AI knows to remove them. 

Getting very annoying - any word on this? Are you aware of the bug at least?

<img width="605" alt="Screenshot 2023-05-11 at 2 51 16 PM" src="https://github.com/community/community/assets/176013/65c1b7e5-4911-4068-8f91-62ab7642ccdc">

**Reply by luke396** [reply]
Can I politely ask, what is the page provided in your screenshot where you can talk (copilot chat maybe)?

**Reply by Stephanie-Spears** [reply]
I'm curious, too

**Reply by wesbos** [reply]
This is Raycast - a desktop app uses GPT-3.5 for chat. Though I am testing Copilot X chat and its even better for this stuff

**Comment by wesbos** [comment]
BTW we had @mattrothenberg on our podcast, and while he doesn't work on Copilot any longer, he said hes aware of the bug too. Hopefully we will hear some sort of confirmation from the team soon.

**Comment by drusepth** [comment]
Just to bump this and confirm this is still an issue -- I'm still seeing this in August, 2023, almost always adding `}` to close all open scopes.

![image](https://github.com/community/community/assets/538235/38d11fdc-3f93-4ad1-bc3f-10dcdf4e3673)

![image](https://github.com/community/community/assets/538235/d012dc86-32c4-49ac-a304-8cab17f66589)

![image](https://github.com/community/community/assets/538235/06307200-34e8-4407-95e5-5c6c66195ab0)

**Comment by pyronaur** [comment]
+1


Side note: This isn't an issue in WebStorm - so it should be doable in VSCode too!

**Reply by Stephanie-Spears** [reply]
Not sure if this is accurate since it was a while ago, but I think I may have mistyped extra quotes on one of the first completion lines I accepted. Had me wondering if all the code suggestions thereafter were using that initial mistake as a baseline...seems unlikely that the AI would get stumped by something like that, though.  ¯\\_(ツ)_/¯

**Comment by isidorn** [comment]
Hi, VS Code PM here 👋 @hediet has fixed this issue 🎉
It should be included in the Copilot extension release in October. Once it is released, I will comment here again so you can try it out and let us know if there are still cases that are missing. fyi @hsenag

Thanks all 🙏

**Reply by b0o** [reply]
Has this fix been released? I've noticed that now Copilot is not closing parens that IMO it should close, e.g:

![2023-10-11_14-13-37_region](https://github.com/community/community/assets/21299126/a8572e05-dfaa-4a49-bed9-408eb277ee39)

I would expect the closing `)` to be included.

**Reply by isidorn** [reply]
Not yet released unfortunately.

**Reply by hsenag** [reply]
Hi, apologies for the delay. @hediet's fix is now in the pre-release version of the extension, **1.127.499** and above. It should make its way into a full release sometime next week.

I would caution that all of this is based on heuristics and it's unlikely we'll ever be able to get this 100% accurate, given that the underlying models are themselves intrinsically unpredictable.

**Reply by wesbos** [reply]
thank you @hsenag! Testing it out now as it seems to have hit stable.

**Reply by KevinVandy** [reply]
is it `v1.130`? I just updated and restarted I still get extra parentheses and ending ``` in mdx

<img width="479" alt="image" src="https://github.com/community/community/assets/28243511/05de0521-37fb-4fbf-81b1-8093c5ab31dd">

Single and double quotes seem to be working better everywhere else though.

**Reply by hsenag** [reply]
> is it v1.130? I just updated and restarted I still get extra parentheses and ending ``` in mdx

Yeah, what we've done so far should be in v1.130. I don't think we'll ever get this perfect given the twin complexities of working with LLMs and unfinished documents, but we'll keep trying to improve common cases.

**Reply by wesbos** [reply]
Still hitting it in CSS

<img width="534" alt="Screenshot 2023-10-26 at 10 19 35 AM" src="https://github.com/community/community/assets/176013/c58bb298-4a1a-4cb6-a273-a574fe663e4e">

**Reply by wesbos** [reply]
Still happening in JS: 
<img width="497" alt="Screenshot 2023-10-26 at 10 27 13 AM" src="https://github.com/community/community/assets/176013/4f13f928-fbf7-4169-aaf6-29e6775d8dbc">

**Reply by wesbos** [reply]
Are we sure the fix is in `v1.130.0` @hsenag? I haven't hit anything yet where it does seem fixed. Even a string:
<img width="415" alt="Screenshot 2023-10-26 at 4 40 16 PM" src="https://github.com/community/community/assets/176013/e72306d2-378c-4208-876f-38e76130988a">

**Reply by hsenag** [reply]
@wesbos We've definitely fixed two different scenarios relating to closing characters being duplicated, and I'm sure the fixes are in `1.130.0`. But I'm also sure there are lots more problematic scenarios, some of which we are already tracking and some of which we aren't.

I had a quick play and I can reproduce your CSS example but not the other two.

**Reply by wesbos** [reply]
@hsenag Seems much better today, can reproduce in an existing file, but not a fresh one. Will try keep notes on situations so I can give you more clear reproductions

**Reply by isidorn** [reply]
@wesbos thanks. Repro steps would be great.

**Reply by hediet** [reply]
The css example is not covered by the bug fix, as the suggestion (`) * 2));`) is not a prefix of text after the suggestion (`)));`).
I think this can be fixed, but I don't know what the implications are for situations where that behavior would be unwanted.

The JS examples look like the bugs we tried to fix though.

**Reply by dnicolson** [reply]
It seems reproducible with `v1.130.0` in a new file using a `const` declaration and then typing `console.log` and pasting the variable name:
![VS Code](https://github.com/community/community/assets/2276355/1ae46869-766a-4b08-911f-1d9382365d35)

Copilot `v1.3.3.3572` on IntelliJ doesn't have the same issue.

**Reply by hsenag** [reply]
@dnicolson this is the suggestion I get in VSCode in that scenario:
<img width="316" alt="image" src="https://github.com/community/community/assets/451141/60a58e30-c7c8-4b50-8c60-39b2c5a919ac">
Not sure what's different.

**Reply by dnicolson** [reply]
Try with this starting code:
```
import crypto from 'crypto';
import { Client } from '@microsoft/microsoft-graph-client';

const FOLDER_PATH = '';
const createShareLink = 'S';
```

![Copilot (1)](https://github.com/community/community/assets/2276355/421b49fa-d852-48d5-9fe1-6e9bcded859f)

**Reply by wesbos** [reply]
Yeah It seems to not happen on fresh files, but as soon as the file has a few lines in it, it starts to happen consistently. In practice, it seems about 20% of the use cases were fixed. Amazon Code Whisperer does not have the same issue since their build `1.97.0 2023-11-02`

**Reply by hsenag** [reply]
@dnicolson thanks - it happens sometimes for me with that example. We'll see what we can do to fix it.

**Reply by wesbos** [reply]
Another two examples. Let me know if I should keep posting them. 

https://github.com/community/community/assets/176013/f4253189-77f1-4e65-bb4b-f662c53d16d0

Code for this one:

```ts
async function getFile(filePath: string) {
  if (cache.has(filePath)) {
    console.log(`[CACHE] ${filePath}`);
    console.log({ cache });
    return cache.get(filePath);
  }

  const file = (await import(filePath)) as { default: string };
  cache.set(filePath, file.default);
  return file.default;
}
```

**Reply by hsenag** [reply]
@wesbos They are useful - I can't promise immediate fixes but I have spent a few hours this afternoon on @dnicolson 's one so they're not just disappearing into the ether.

**Reply by ian-h-chamberlain** [reply]
I've run into this case somewhat frequently with Rust `#[attributes]`, particularly with parentheses. Example:

https://github.com/rust3ds/citro3d-rs/blob/c56c4cdfed331355d0f4cf305487116dcfa25a37/citro3d/src/lib.rs#L121

<img width="601" alt="Screen Shot 2023-11-24 at 9 04 36 PM" src="https://github.com/community/community/assets/11131775/c426dc01-fbe1-4997-be74-e1f9a4ca6385">

In that case `#[doc()]` brackets were completed by VSCode, but Copilot still suggested an additional parenthesis. This seems to happen maybe 50-75% of the time for me.

Copilot version `v1.138.0`, in case that's relevant. Hope these reports are still useful.

**Comment by wesbos** [comment]
Still seems to be an issue, and maybe reverted? See it with console.log and most methods that need a string

<img width="306" alt="Screenshot 2024-01-10 at 12 17 26 PM" src="https://github.com/community/community/assets/176013/4170b8db-0608-402e-815a-a6e5cfed5c2e">
<img width="320" alt="Screenshot 2024-01-10 at 12 18 31 PM" src="https://github.com/community/community/assets/176013/d808b94b-fcd8-4a9e-aaef-441b0d9e9c5a">
<img width="460" alt="Screenshot 2024-01-10 at 12 19 03 PM" src="https://github.com/community/community/assets/176013/9067cba9-01d3-4e00-9cca-3c7289917c50">
<img width="280" alt="Screenshot 2024-01-10 at 12 20 46 PM" src="https://github.com/community/community/assets/176013/127a3e28-22cd-4d1e-9d5a-7a6e7eb7a6af">

**Comment by Superskyyy** [comment]
Still a major ux issue after a year :(

**Comment by xixixao** [comment]
Isn't there a simple, but imperfect fix? Take the suggestion from copilot, compare it to what's after the cursor, and strip out anything that matches from the end.

There will be scenarios in which too much is stripped, especially if the punctuation is already imbalanced, but I think those will be rarer and easier to fix than the super annoying status quo.

@wesbos (thank you for fighting the good fight), @hsenag @hediet @isidorn @idan 

Bonus points: Sometimes copilot also suggests trailing semicolon which is usually not yet typed, so consider that being present if there's a new line or end of file.

**Comment by tpope** [comment, marked as answer]
Pre-release 1.152.662 should contain fixes to most if not all of the duplicate bracket/quote bugs reported in this discussion.

**Reply by wesbos** [reply]
I think it might be fixed?! Im gonna keep testing but I can't get it to goof up! Thank you
<img width="607" alt="Screenshot 2024-01-19 at 2 07 31 PM" src="https://github.com/community/community/assets/176013/7f2f93b0-3e7e-43bc-9811-e9e7190838ed">
<img width="460" alt="Screenshot 2024-01-19 at 2 04 56 PM" src="https://github.com/community/community/assets/176013/17867d05-18a3-4652-9768-2b32b8fbfabe">
<img width="393" alt="Screenshot 2024-01-19 at 2 04 38 PM" src="https://github.com/community/community/assets/176013/62b829a7-8da6-4d14-9801-a777c19a054a">

**Reply by wesbos** [reply]
Not sure if this was covered, but multi-line backtick strings still add an extra 

<img width="389" alt="Screenshot 2024-01-19 at 2 57 13 PM" src="https://github.com/community/community/assets/176013/0f3a619d-e49d-47d0-a0f8-e0014a65db01">

**Reply by ajwootto** [reply]
I experience a lot of these issues in Jetbrains IDE as well. Is it also going to be fixed there?

**Reply by tpope** [reply]
The big fix from last week was common to all plugins and has landed in stable releases of Copilot for VS Code (1.153.0+), JetBrains (1.4.12.4386+), and Vim/Neovim (1.16.0+).

~~There's still a few multi-line edge cases; hopefully will have more to report soon.~~ :heavy_check_mark:

**Reply by wesbos** [reply]
Thank you @tpope! i've been very happy with it since this. Ill mark as fixed!

**Reply by tpope** [reply]
Pre-release 1.155.677 has a fix for multi-line issues like the one with the JavaScript backtick string above. This too will make its way into upcoming stable releases of all editor plugins.
